File: VB/GrowthWare.Framework.BusinessData/DataAccessLayer/SQLServer/V2008/DFunctions.vb
Imports System.Data.SqlClient
Imports GrowthWare.Framework.Model.Profiles
Imports GrowthWare.Framework.BusinessData.DataAccessLayer.Interfaces
Imports GrowthWare.Framework.BusinessData.DataAccessLayer.SQLServer.Base
Imports GrowthWare.Framework.Model.Enumerations

Namespace DataAccessLayer.SQLServer.V2008
    Public Class DFunctions
        Inherits DDBInteraction
        Implements IDFunction

#Region "Member Objects"
        Private m_Profile As MFunctionProfile = Nothing
#End Region

#Region "Public Properties"
        Public ReadOnly Property GetFunction As System.Data.DataRow Implements IDFunction.GetFunction
            Get
                checkValid()
                Dim mParameters() As SqlParameter =
                {
                 New SqlParameter("@P_Function_SeqID", Profile.Id)
                }
                Return MyBase.GetDataRow("ZGWSecurity.Get_Function", mParameters)
            End Get
        End Property

        Public ReadOnly Property GetFunctions As System.Data.DataSet Implements IDFunction.GetFunctions
            Get
                checkValid()
                Dim mDSFunctions As DataSet = Nothing
                Dim mParameters() As SqlParameter =
                {
                 New SqlParameter("@P_Function_SeqID", m_Profile.Id)
                }
                Try
                    Dim mFunctions As DataTable = MyBase.GetDataTable("ZGWSecurity.Get_Function", mParameters)
                    mDSFunctions = Me.GetSecurity()
                    mDSFunctions.Tables(0).TableName = "DerivedRoles"
                    mDSFunctions.Tables(1).TableName = "AssignedRoles"
                    mDSFunctions.Tables(2).TableName = "Groups"


                    Dim mHasAssingedRoles As Boolean = False
                    Dim mHasGroups As Boolean = False
                    mFunctions.TableName = "Functions"
                    If mDSFunctions.Tables("AssignedRoles").Rows.Count > 0 Then mHasAssingedRoles = True
                    If mDSFunctions.Tables("Groups").Rows.Count > 0 Then mHasGroups = True
                    mDSFunctions.Tables.Add(mFunctions.Copy)

                    Dim mRelation As DataRelation = New DataRelation("DerivedRoles", mDSFunctions.Tables("Functions").Columns("Function_Seq_ID"), mDSFunctions.Tables("DerivedRoles").Columns("Function_Seq_ID"))
                    mDSFunctions.Relations.Add(mRelation)
                    If mHasAssingedRoles Then
                        mRelation = New DataRelation("AssignedRoles", mDSFunctions.Tables("Functions").Columns("Function_Seq_ID"), mDSFunctions.Tables("AssignedRoles").Columns("Function_Seq_ID"))
                        mDSFunctions.Relations.Add(mRelation)
                    End If
                    If mHasGroups Then
                        mRelation = New DataRelation("Groups", mDSFunctions.Tables("Functions").Columns("Function_Seq_ID"), mDSFunctions.Tables("Groups").Columns("Function_Seq_ID"))
                        mDSFunctions.Relations.Add(mRelation)
                    End If
                Catch ex As Exception
                    If Not mDSFunctions Is Nothing Then
                        mDSFunctions.Dispose()
                    End If
                    Throw
                End Try
                Return mDSFunctions
            End Get
        End Property

        Public Property Profile As MFunctionProfile Implements IDFunction.Profile
            Get
                Return m_Profile
            End Get
            Set(value As MFunctionProfile)
                m_Profile = value
            End Set
        End Property

        Public Property SecurityEntitySeqID As Integer Implements IDFunction.SecurityEntitySeqID
#End Region

#Region "Pulbic Methods"
        ''' <summary>
        ''' Deletes an account
        ''' </summary>
        Public Sub Delete(ByVal functionSeqId As Integer) Implements IDFunction.Delete
            Dim myParameters() As SqlParameter =
            {
              New SqlParameter("@P_Function_SeqID", functionSeqId), _
              GetSqlParameter("@P_ErrorCode", "", ParameterDirection.Output)
            }
            MyBase.ExecuteNonQuery("ZGWSecurity.Delete_Function", myParameters)
        End Sub

        Public Function GetFunctionTypes() As DataTable Implements IDFunction.GetFunctionTypes
            Dim mStoreProcedure As String = "ZGWSecurity.Get_Function_Types"
            Dim mParameters() As SqlParameter = {New SqlParameter("@P_Function_Type_SeqID", -1)}
            Return MyBase.GetDataTable(mStoreProcedure, mParameters)
        End Function

        Public Function GetMenuOrder(ByRef profile As MFunctionProfile) As DataTable Implements IDFunction.GetMenuOrder
            Dim mStoreProcedure As String = "ZGWSecurity.Get_Function_Sort"
            Dim mParameters() As SqlParameter = {New SqlParameter("@P_Function_SeqID", profile.Id)}
            Return MyBase.GetDataTable(mStoreProcedure, mParameters)
        End Function

        ''' <summary>
        ''' Inserts or updates account information
        ''' </summary>
        Public Function Save() As Integer Implements IDFunction.Save
            checkValid()
            Dim mRetVal As Integer = -1
            Dim mParameters() As SqlParameter =
             {
              GetSqlParameter("@P_Function_SeqID", m_Profile.Id, ParameterDirection.InputOutput),
              New SqlParameter("@P_Name", m_Profile.Name),
              New SqlParameter("@P_Description", m_Profile.Description),
              New SqlParameter("@P_Function_Type_SeqID", m_Profile.FunctionTypeSeqId),
              New SqlParameter("@P_Source", m_Profile.Source),
              New SqlParameter("@P_Enable_View_State", m_Profile.EnableViewState),
              New SqlParameter("@P_Enable_Notifications", m_Profile.EnableNotifications),
              New SqlParameter("@P_Redirect_On_Timeout", m_Profile.RedirectOnTimeout),
              New SqlParameter("@P_IS_NAV", m_Profile.IsNavigable),
              New SqlParameter("P_Link_Behavior", m_Profile.LinkBehavior),
              New SqlParameter("@P_NO_UI", m_Profile.NoUI),
              New SqlParameter("@P_NAV_TYPE_ID", m_Profile.NavigationTypeSeqId),
              New SqlParameter("@P_Action", m_Profile.Action),
              New SqlParameter("@P_Meta_Key_Words", m_Profile.MetaKeywords),
              New SqlParameter("@P_Parent_SeqID", m_Profile.ParentId),
              New SqlParameter("@P_Notes", m_Profile.Notes),
              New SqlParameter("@P_Added_Updated_By", GetAddedUpdatedBy(m_Profile))
             }
            MyBase.ExecuteNonQuery("ZGWSecurity.Set_Function", mParameters)
            mRetVal = Integer.Parse(GetParameterValue("@P_Function_SeqID", mParameters))
            Return mRetVal
        End Function

        ''' <summary>
        ''' Save groups by passing a string or comma seporated groups to the database.
        ''' </summary>
        ''' <param name="permission">The permission.</param>
        Public Sub SaveGroups(ByVal permission As PermissionType) Implements IDFunction.SaveGroups
            checkValid()
            Dim mCommaSeporatedString As String = m_Profile.GetCommaSeparatedGroups(permission)
            Dim mParameters() As SqlParameter =
             {
              New SqlParameter("@P_Function_SeqID", Profile.Id),
              New SqlParameter("@P_Security_Entity_SeqID", SecurityEntitySeqID),
              New SqlParameter("@P_Groups", mCommaSeporatedString),
              New SqlParameter("@P_Permissions_NVP_Detail_SeqID", permission),
              New SqlParameter("@P_Added_Updated_By", GetAddedUpdatedBy(Profile))
             }
            MyBase.ExecuteNonQuery("ZGWSecurity.Set_Function_Groups", mParameters)
        End Sub

        ''' <summary>
        ''' Save roles by passing the permission type and sending the profiles comma separated roles to the database.
        ''' </summary>
        ''' <param name="permission">The permission.</param>
        Public Sub SaveRoles(ByVal permission As PermissionType) Implements IDFunction.SaveRoles
            checkValid()
            Dim mCommaSeporatedString As String = Profile.GetCommaSeparatedAssignedRoles(permission)
            Dim mParameters() As SqlParameter =
             {
              New SqlParameter("@P_Function_SeqID", Profile.Id),
              New SqlParameter("@P_Security_Entity_SeqID", SecurityEntitySeqID),
              New SqlParameter("@P_Roles", mCommaSeporatedString),
              New SqlParameter("@P_Permissions_NVP_Detail_SeqID", permission),
              New SqlParameter("@P_Added_Updated_By", GetAddedUpdatedBy(Profile))
             }
            MyBase.ExecuteNonQuery("ZGWSecurity.Set_Function_Roles", mParameters)
        End Sub

        ''' <summary>
        ''' Returns a data table based on the search criteria.
        ''' </summary>
        ''' <param name="searchCriteria">MSearchCriteria</param>
        ''' <returns>DataTable</returns>
        Public Function Search(ByVal searchCriteria As MSearchCriteria) As DataTable Implements IDFunction.Search
            Dim mRetVal As DataTable
            Dim mParameters() As SqlParameter =
             {
              New SqlParameter("@P_Columns", searchCriteria.Columns),
              New SqlParameter("@P_OrderByColumn", searchCriteria.OrderByColumn),
              New SqlParameter("@P_OrderByDirection", searchCriteria.OrderByDirection),
              New SqlParameter("@P_PageSize", searchCriteria.PageSize),
              New SqlParameter("@P_SelectedPage", searchCriteria.SelectedPage),
              New SqlParameter("@P_TableOrView", "ZGWSystem.vwSearchFunctions"),
              New SqlParameter("@P_WhereClause", searchCriteria.WhereClause)
             }
            mRetVal = MyBase.GetDataTable("ZGWSystem.Get_Paginated_Data", mParameters)
            Return mRetVal
        End Function
#End Region

#Region "Private Methods"
        Private Sub checkValid()
            MyBase.IsValid()
            If Profile Is Nothing Then
                Throw New ArgumentException("Profile property must be set before calling methods from this class")
            End If
            If SecurityEntitySeqID = 0 Then
                Throw New ArgumentException("SE_SEQ_ID property must be set before calling methods from this class")
            End If
        End Sub

        Private Function GetSecurity() As System.Data.DataSet
            Dim myStoreProcedure As String = "ZGWSecurity.Get_Function_Security"
            Dim myParameters() As SqlParameter =
            {
             New SqlParameter("@P_Security_Entity_SeqID", SecurityEntitySeqID)
            }
            Return MyBase.GetDataSet(myStoreProcedure, myParameters)
        End Function
#End Region
    End Class
End Namespace

File: Net_Core/DataAccess/SQLServer/DFunctions.cs
using GrowthWare.DataAccess.Interfaces;
using GrowthWare.DataAccess.SQLServer.Base;
using GrowthWare.Framework.Enumerations;
using GrowthWare.Framework.Models;
using System;
using System.Data;
using System.Data.SqlClient;
using System.Globalization;
using System.Threading.Tasks;

namespace GrowthWare.DataAccess.SQLServer;

/// <summary>
/// DFunctions provides all database interaction to SQL Server 2008
/// </summary>
/// <remarks>
///		The Profile and SecurityEntitySeqID properties must be set
///		before using any methods.
///		Properties where chosen instead of parameters because all
///		methods will need one or both to perform their work.
///	</remarks>
public class DFunctions : AbstractDBInteraction, IFunction
{

    #region Member Fields
    private MFunctionProfile m_Profile = null;
    private int m_SecurityEntitySeqId = -2;
    #endregion

    #region Public Properties
    async Task<DataRow> IFunction.GetFunction()
    {
        checkValid();
        SqlParameter[] mParameters = [
            new("@P_FunctionSeqId", m_Profile.Id)
        ];
        String mStoreProcedure = "[ZGWSecurity].[Get_Function]";
        return await base.GetDataRowAsync(mStoreProcedure, mParameters);
    }

    async Task<DataSet> IFunction.GetFunctions(int securityEntitySeqId)
    {
        DataSet mDSFunctions = null;
        checkValid();
        SqlParameter[] mParameters = [
              new("@P_FunctionSeqId", m_Profile.Id)
            , new("@P_SecurityEntitySeqId", securityEntitySeqId)
        ];
        try
        {
            string mStoredProcedure = "[ZGWSecurity].[Get_Function]";
            mDSFunctions = await this.GetDataSetAsync(mStoredProcedure, mParameters);
            mDSFunctions.Tables[(int)FunctionSecurityTables.DerivedRoles].TableName = FunctionSecurityTableNames.DERIVED_ROLES;
            mDSFunctions.Tables[(int)FunctionSecurityTables.AssignedRoles].TableName = FunctionSecurityTableNames.ASSIGNED_ROLES;
            mDSFunctions.Tables[(int)FunctionSecurityTables.AssignedGroups].TableName = FunctionSecurityTableNames.ASSIGNED_GROUPS;
            mDSFunctions.Tables[(int)FunctionSecurityTables.Functions].TableName = FunctionSecurityTableNames.FUNCTIONS;


            bool mHasAssingedRoles = false;
            bool mHasGroups = false;
            if (mDSFunctions.Tables[FunctionSecurityTableNames.ASSIGNED_ROLES].Rows.Count > 0) mHasAssingedRoles = true;
            if (mDSFunctions.Tables[FunctionSecurityTableNames.ASSIGNED_GROUPS].Rows.Count > 0) mHasGroups = true;

            DataRelation mRelation = new DataRelation(FunctionSecurityTableNames.DERIVED_ROLES, mDSFunctions.Tables[FunctionSecurityTableNames.FUNCTIONS].Columns["Function_Seq_ID"], mDSFunctions.Tables[FunctionSecurityTableNames.DERIVED_ROLES].Columns["Function_Seq_ID"]);
            mDSFunctions.Relations.Add(mRelation);
            if (mHasAssingedRoles)
            {
                mRelation = new DataRelation(FunctionSecurityTableNames.ASSIGNED_ROLES, mDSFunctions.Tables[FunctionSecurityTableNames.FUNCTIONS].Columns["Function_Seq_ID"], mDSFunctions.Tables[FunctionSecurityTableNames.ASSIGNED_GROUPS].Columns["Function_Seq_ID"]);
                mDSFunctions.Relations.Add(mRelation);
            }
            if (mHasGroups)
            {
                mRelation = new DataRelation(FunctionSecurityTableNames.ASSIGNED_GROUPS, mDSFunctions.Tables[FunctionSecurityTableNames.FUNCTIONS].Columns["Function_Seq_ID"], mDSFunctions.Tables[FunctionSecurityTableNames.ASSIGNED_GROUPS].Columns["Function_Seq_ID"]);
                mDSFunctions.Relations.Add(mRelation);
            }

        }
        catch (Exception)
        {

            throw;
        }
        return mDSFunctions;
    }

    async Task<DataTable> IFunction.MenuTypes()
    {
        string mStoreProcedure = "[ZGWSecurity].[Get_Menu_Types]";
        SqlParameter[] mParameters = [new("@P_FunctionTypeSeqId", -1)];
        return await base.GetDataTableAsync(mStoreProcedure, mParameters);
    }

    MFunctionProfile IFunction.Profile
    {
        get
        {
            return m_Profile;
        }
        set
        {
            m_Profile = value;
        }
    }

    int IFunction.SecurityEntitySeqId
    {
        get
        {
            return m_SecurityEntitySeqId;
        }
        set
        {
            m_SecurityEntitySeqId = value;
        }
    }
    #endregion

    #region Constructors
    public DFunctions(string connectionString, int securityEntitySeqId) : base()
    {
        this.ConnectionString = connectionString;
        this.m_SecurityEntitySeqId = securityEntitySeqId;
    }
    #endregion

    async Task IFunction.CopyFunctionSecurity(int source, int target, int added_Updated_By)
    {
        SqlParameter[] mParameters = [
            new("@P_Source", source),
            new("@P_Target", target),
            new("@P_Added_Updated_By", added_Updated_By)
        ];
        String mStoreProcedure = "[ZGWSecurity].[Copy_Function_Security]";
        await base.ExecuteNonQueryAsync(mStoreProcedure, mParameters);
    }
    async Task IFunction.Delete(int functionSeqId)
    {
        SqlParameter[] mParameters = [
            new("@P_FunctionSeqId", functionSeqId),
            GetSqlParameter("@P_ErrorCode", "", ParameterDirection.Output)
        ];
        String mStoreProcedure = "[ZGWSecurity].[Delete_Function]";
        await base.ExecuteNonQueryAsync(mStoreProcedure, mParameters);
    }

    async Task<DataTable> IFunction.FunctionTypes()
    {
        string mStoreProcedure = "[ZGWSecurity].[Get_Function_Types]";
        SqlParameter[] mParameters = [new("@P_FunctionTypeSeqId", -1)];
        return await base.GetDataTableAsync(mStoreProcedure, mParameters);
    }

    async Task<DataTable> IFunction.GetMenuOrder(MFunctionProfile Profile)
    {
        string mStoreProcedure = "[ZGWSecurity].[Get_Function_Sort]";
        SqlParameter[] mParameters = [new("@P_FunctionSeqId", Profile.Id)];
        return await base.GetDataTableAsync(mStoreProcedure, mParameters);
    }

    async Task<int> IFunction.Save()
    {
        SqlParameter[] mParameters = [
            GetSqlParameter("@P_FunctionSeqId", m_Profile.Id, ParameterDirection.InputOutput),
            new("@P_Name", m_Profile.Name),
            new("@P_Description", m_Profile.Description ?? ""),
            new("@P_FunctionTypeSeqId", m_Profile.FunctionTypeSeqId),
            new("@P_Source", m_Profile.Source ?? ""),
            new("@P_Controller", m_Profile.Controller ?? ""),
            new("@P_Enable_View_State", m_Profile.EnableViewState),
            new("@P_Enable_Notifications", m_Profile.EnableNotifications),
            new("@P_Redirect_On_Timeout", m_Profile.RedirectOnTimeout),
            new("@P_IS_NAV", m_Profile.IsNavigable),
            new("@P_Link_Behavior", m_Profile.LinkBehavior),
            new("@P_NO_UI", m_Profile.NoUI),
            new("@P_NAV_TYPE_ID", m_Profile.NavigationTypeSeqId),
            new("@P_Action", m_Profile.Action),
            new("@P_Meta_Key_Words", m_Profile.MetaKeywords ?? ""),
            new("@P_ParentSeqId", m_Profile.ParentId),
            new("@P_Notes", m_Profile.Notes ?? ""),
            new("@P_Added_Updated_By", GetAddedUpdatedBy(m_Profile, m_Profile.Id))
        ];
        String mStoreProc = "[ZGWSecurity].[Set_Function]";
        await base.ExecuteNonQueryAsync(mStoreProc, mParameters);
        return int.Parse(GetParameterValue("@P_FunctionSeqId", mParameters), CultureInfo.InvariantCulture);
    }

    async Task IFunction.SaveGroups(PermissionType permission, int securityEntitySeqId)
    {
        checkValid();
        String mCommaSeporatedString = m_Profile.GetCommaSeparatedGroups(permission);
        string mStoreProcedure = "[ZGWSecurity].[Set_Function_Groups]";
        SqlParameter[] mParameters = [
            new("@P_FunctionSeqId", m_Profile.Id),
            new("@P_SecurityEntitySeqId", securityEntitySeqId),
            new("@P_Groups", mCommaSeporatedString),
            new("@P_PermissionsNVPDetailSeqId", permission),
            new("@P_Added_Updated_By", GetAddedUpdatedBy(m_Profile, m_Profile.Id))
        ];
        await base.ExecuteNonQueryAsync(mStoreProcedure, mParameters);
    }

    async Task IFunction.SaveRoles(PermissionType permission, int securityEntitySeqId)
    {
        checkValid();
        String mCommaSeporatedString = m_Profile.GetCommaSeparatedAssignedRoles(permission);
        string mStoreProcedure = "[ZGWSecurity].[Set_Function_Roles]";
        SqlParameter[] mParameters = [
            new("@P_FunctionSeqId", m_Profile.Id),
            new("@P_SecurityEntitySeqId", securityEntitySeqId),
            new("@P_Roles", mCommaSeporatedString),
            new("@P_PermissionsNVPDetailSeqId", permission),
            new("@P_Added_Updated_By", GetAddedUpdatedBy(m_Profile, m_Profile.Id))
        ];
        await base.ExecuteNonQueryAsync(mStoreProcedure, mParameters);
    }

    async Task IFunction.UpdateMenuOrder(string commaSeparated_Ids, MFunctionProfile profile)
    {
        string mStoreProcedure = "[ZGWSecurity].[Set_Function_Sort]";
        SqlParameter[] mParameters = [
            new("@P_Commaseparated_Ids", commaSeparated_Ids),
            new("@P_Added_Updated_By", GetAddedUpdatedBy(profile, m_Profile.Id)),
            GetSqlParameter("@P_Primary_Key", "", ParameterDirection.Output)
        ];
        await base.ExecuteNonQueryAsync(mStoreProcedure, mParameters);
    }

    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA2204:Literals should be spelled correctly", MessageId = "SecurityEntitySeqId")]
    private void checkValid()
    {
        base.IsValid();
        if (m_Profile == null)
        {
            throw new InvalidOperationException("The Profile property must set before using any methods from this class.");
        }
        if (m_SecurityEntitySeqId == 0)
        {
            throw new InvalidOperationException("The SecurityEntitySeqId property must set before using any methods from this class.");
        }
    }

    private async Task<DataSet> getSecurity()
    {
        string mStoreProcedure = "[ZGWSecurity].[Get_Function_Security]";
        SqlParameter[] mParameters = [new("@P_SecurityEntitySeqId", m_SecurityEntitySeqId)];
        return await base.GetDataSetAsync(mStoreProcedure, mParameters);
    }
}
